File: NativeApp/ViewModels/OllamaPayViewModel.swift
import Foundation
import SwiftUI
import AppKit

// MARK: - Ollama Pay ViewModel
@MainActor
class OllamaPayViewModel: ObservableObject {
    @Published var apiKey: String {
        didSet {
            let cleanKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
            if cleanKey.isEmpty {
                KeychainHelper.delete(key: "OllamaPayAPIKey")
            } else {
                KeychainHelper.save(key: "OllamaPayAPIKey", data: cleanKey)
            }
        }
    }
    @Published var refreshInterval: Int {
        didSet {
            UserDefaults.standard.set(refreshInterval, forKey: "OllamaPayRefreshInterval")
            startAutoRefresh()
        }
    }
    @Published var isAvailable: Bool = true
    @Published var isLoading: Bool = false
    @Published var todayTokens: Int = 0
    @Published var totalTokensUsed: Int = 0
    @Published var tokensLimit: Int = 0
    @Published var errorMsg: String = ""
    @Published var monthTokens: Int = 0
    @Published var monthRequests: Int = 0
    @Published var todayRequests: Int = 0
    @Published var tokensRemaining: Int = 0
    @Published var todayPloyJoyTokens: Int = 0
    @Published var latencyMs: Int = 0
    
    private var refreshTimer: Timer?
    
    init() {
        self.apiKey = KeychainHelper.load(key: "OllamaPayAPIKey") ?? ""
        let savedInterval = UserDefaults.standard.integer(forKey: "OllamaPayRefreshInterval")
        self.refreshInterval = savedInterval > 0 ? savedInterval : 60
        fetchData()
        startAutoRefresh()
    }
    
    func startAutoRefresh() {
        refreshTimer?.invalidate()
        guard refreshInterval > 0 else { return }
        refreshTimer = Timer.scheduledTimer(withTimeInterval: TimeInterval(refreshInterval), repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.fetchData()
            }
        }
    }
    
    func fetchData() {
        let cleanKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleanKey.isEmpty, let url = URL(string: APIConstants.OllamaPay.usageTotalURL) else {
            self.isAvailable = false
            self.errorMsg = "API Key not set"
            AppDelegate.shared?.updateOllamaPayTitle(todayStr: "No Key")
            return
        }
        
        self.isLoading = true
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(cleanKey)", forHTTPHeaderField: "Authorization")
        
        let start = Date()
        
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                self.isLoading = false
                self.latencyMs = Int(Date().timeIntervalSince(start) * 1000)
                if let error = error {
                    self.isAvailable = false
                    self.errorMsg = error.localizedDescription
                    return
                }
                guard let data = data else {
                    self.isAvailable = false
                    self.errorMsg = "No data received"
                    return
                }
                
                if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
                    self.isAvailable = false
                    self.errorMsg = "Error \(httpResponse.statusCode)"
                    AppDelegate.shared?.updateOllamaPayTitle(todayStr: "Err")
                    
                    if httpResponse.statusCode == 429 {
                        NotificationService.shared.sendNotification(
                            identifier: "ollama_pay_rate_limit",
                            title: "⚠️ Ollama Pay Rate Limited",
                            body: "Rate limit reached for Ollama Pay API."
                        )
                    }
                    return
                }
                
                do {
                    let res = try JSONDecoder().decode(OllamaPayUsageResponse.self, from: data)
                    self.isAvailable = true
                    self.todayTokens = res.accounting?.todayTokens ?? 0
                    self.totalTokensUsed = res.tokensUsed ?? 0
                    self.tokensLimit = res.tokensLimit ?? 0
                    self.errorMsg = ""
                    self.monthTokens = res.accounting?.monthTokens ?? 0
                    self.monthRequests = res.accounting?.monthRequests ?? 0
                    self.todayRequests = res.accounting?.todayRequests ?? 0
                    self.tokensRemaining = res.tokensRemaining ?? 0
                    self.todayPloyJoyTokens = res.accounting?.todayPloyJoyTokens ?? 0
                    
                    AppDelegate.shared?.updateOllamaPayTitle(todayStr: self.todayTokens.formattedTokenString)
                } catch {
                    self.isAvailable = false
                    self.errorMsg = "Parse error"
                    AppDelegate.shared?.updateOllamaPayTitle(todayStr: "Err")
                }
            }
        }.resume()
    }
    
    func saveSettings() {
        startAutoRefresh()
        fetchData()
    }
}

File: WindowsApp/AIWidgetWindowsApp/ViewModels/OllamaPayViewModel.cs
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Threading.Tasks;
using System.Windows;
using AIWidgetWindowsApp.Helpers;
using AIWidgetWindowsApp.Models;
using AIWidgetWindowsApp.Services;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;

namespace AIWidgetWindowsApp.ViewModels
{
    public partial class OllamaPayViewModel : ObservableObject
    {
        private readonly HttpClient _httpClient = new HttpClient();

        [ObservableProperty]
        private string apiKey = string.Empty;

        partial void OnApiKeyChanged(string value)
        {
            KeychainHelper.Save("OllamaPayAPIKey", value);
        }

        private System.Windows.Threading.DispatcherTimer? _refreshTimer;

        [ObservableProperty]
        private int refreshInterval = 60;

        partial void OnRefreshIntervalChanged(int value)
        {
            KeychainHelper.Save("OllamaPayRefreshInterval", value.ToString());
            StartAutoRefresh();
        }

        [ObservableProperty]
        private bool isAvailable = true;

        [ObservableProperty]
        private bool isLoading = false;

        [ObservableProperty]
        private int todayTokens = 0;

        [ObservableProperty]
        private int totalTokensUsed = 0;

        [ObservableProperty]
        private int tokensLimit = 0;

        [ObservableProperty]
        private string errorMsg = string.Empty;

        [ObservableProperty]
        private int monthTokens = 0;

        [ObservableProperty]
        private int monthRequests = 0;

        [ObservableProperty]
        private int todayRequests = 0;

        [ObservableProperty]
        private int tokensRemaining = 0;

        [ObservableProperty]
        private int todayPloyJoyTokens = 0;

        [ObservableProperty]
        private int latencyMs = 0;

        public OllamaPayViewModel()
        {
            this.ApiKey = KeychainHelper.Load("OllamaPayAPIKey") ?? string.Empty;
            
            if (int.TryParse(KeychainHelper.Load("OllamaPayRefreshInterval"), out int savedInterval) && savedInterval > 0)
            {
                this.RefreshInterval = savedInterval;
            }
            
            _ = FetchDataAsync();
            StartAutoRefresh();
        }

        public void StartAutoRefresh()
        {
            _refreshTimer?.Stop();
            if (RefreshInterval > 0)
            {
                _refreshTimer = new System.Windows.Threading.DispatcherTimer();
                _refreshTimer.Interval = TimeSpan.FromSeconds(RefreshInterval);
                _refreshTimer.Tick += (s, e) => _ = FetchDataAsync();
                _refreshTimer.Start();
            }
        }

        [RelayCommand]
        public async Task FetchDataAsync()
        {
            var cleanKey = ApiKey?.Trim();
            if (string.IsNullOrWhiteSpace(cleanKey))
            {
                IsAvailable = false;
                ErrorMsg = "API Key not set";
                UpdateMainWindow("No Key");
                return;
            }

            IsLoading = true;
            var stopwatch = Stopwatch.StartNew();

            try
            {
                var request = new HttpRequestMessage(HttpMethod.Get, APIConstants.OllamaPay.UsageTotalURL);
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", cleanKey);

                var response = await _httpClient.SendAsync(request);
                stopwatch.Stop();
                LatencyMs = (int)stopwatch.ElapsedMilliseconds;

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    var res = JsonSerializer.Deserialize<OllamaPayUsageResponse>(json);

                    if (res != null)
                    {
                        IsAvailable = true;
                        ErrorMsg = string.Empty;
                        TodayTokens = res.Accounting?.TodayTokens ?? 0;
                        TotalTokensUsed = res.TokensUsed ?? 0;
                        TokensLimit = res.TokensLimit ?? 0;
                        MonthTokens = res.Accounting?.MonthTokens ?? 0;
                        MonthRequests = res.Accounting?.MonthRequests ?? 0;
                        TodayRequests = res.Accounting?.TodayRequests ?? 0;
                        TokensRemaining = res.TokensRemaining ?? 0;
                        TodayPloyJoyTokens = res.Accounting?.TodayPloyJoyTokens ?? 0;

                        UpdateMainWindow(TodayTokens.ToString("N0") + " tokens");
                    }
                }
                else
                {
                    IsAvailable = false;
                    ErrorMsg = $"Error {(int)response.StatusCode}";
                    UpdateMainWindow("Err");

                    if (response.StatusCode == System.Net.HttpStatusCode.TooManyRequests)
                    {
                        NotificationService.SendNotification("ollama_pay_rate_limit", "⚠️ Ollama Pay Rate Limited", "Rate limit reached for Ollama Pay API.");
                    }
                }
            }
            catch (Exception ex)
            {
                IsAvailable = false;
                ErrorMsg = ex.Message;
                UpdateMainWindow("Err");
            }
            finally
            {
                IsLoading = false;
            }
        }

        private void UpdateMainWindow(string value)
        {
            if (Application.Current.MainWindow is MainWindow mainWindow)
            {
                mainWindow.UpdateTrayTextIfSelected(3, value);
            }
        }
    }
}
